File: CountHillsAVnnA.py
from typing import List

class Solution:
    def countHillValley(self, nums: List[int]) -> int:
        count = 0
        # Remove consecutive duplicates to simplify comparisons
        filtered = [nums[0]]
        for num in nums[1:]:
            if num != filtered[-1]:
                filtered.append(num)

        # Check each middle element for hill or valley
        for i in range(1, len(filtered) - 1):
            if filtered[i] > filtered[i - 1] and filtered[i] > filtered[i + 1]:
                count += 1  # Hill
            elif filtered[i] < filtered[i - 1] and filtered[i] < filtered[i + 1]:
                count += 1  # Valley

        return count

File: CountHillsAVnnA.cpp
class Solution {
public:
    int countHillValley(std::vector<int>& nums) {
        int count = 0;
        int n = nums.size();

        // Track the last distinct value to the left
        int left = nums[0];

        for (int i = 1; i + 1 < n; ++i) {
            // Skip if current is equal to next (part of same terrain)
            if (nums[i] == nums[i + 1]) continue;

            // Check for hill or valley
            if ((nums[i] > left && nums[i] > nums[i + 1]) ||
                (nums[i] < left && nums[i] < nums[i + 1])) {
                ++count;
            }

            // Update left only when current is not equal to next
            left = nums[i];
        }

        return count;
    }
};
